# DEG Analysis
## Introduction
In this analysis we identifed genes that are differentially expressed based on the microarray results of there expressions. These genes are expressed by cells that infected by 
1. a wild type sars-cov virus, icSARs, 
2. dORF-6 mutant sars-cov, 
3. BAT mutant virus and 
4. H1N1, influenza virus.

The data used for this study were extracted from the work of Mitchell et al (2013).

Data access on this [Link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE47960)

## Methods
The defferential gene expression (DEG) analysis can be found [here](https://github.com/d1stadeyemi/RNA_seq/blob/master/gene_exp.ipynb)
![Methodology](https://github.com/d1stadeyemi/RNA_seq/blob/master/images/methodology.png)

## Results
1. We identified 25 common DEGs across influenza and SARS-CoV infections, highlighting shared host responses.
2. CXCL10, CXCL2, and CXCL3 are central to chemokine activity and CXCR receptor binding, driving immune cell recruitment.
3. FOS, JUNB, ZFP36, DUSP1, and EGR1 regulate inflammatory responses, linking viral infection to immune modulation.
4. HEXDC, XAF1, and MAP9 are involved in cellular stress and structural integrity, suggesting potential roles in viral pathogenesis.
5. These findings reveal key immune regulators, providing insights into therapeutic targets for controlling viral-induced inflammation.

## Conclusion
We uncovered shared immune responses across influenza and SARS-CoV infections highlights key targets for therapeutic intervention. This study showed that chemokine signaling drives immune recruitment, potentially influencing disease severity and inflammation. Also, regulators of inflammation and cellular stress may contribute to viral pathogenesis and host recovery.
These insights provide a foundation for future studies on host-directed therapies and viral immune modulation.